pkg/authorization/admission/restrictusers/subjectchecker.go

```diff
@@ -0,0 +1,181 @@
+package restrictusers
+
+import (
+	"fmt"
+
+	kclientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+
+	kapi "k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/labels"
+
+	authorizationapi "github.com/openshift/origin/pkg/authorization/api"
+	oclient "github.com/openshift/origin/pkg/client"
+)
+
+// SubjectChecker determines whether a subject (user, group, or service account)
+// is a collaborator on a project.  A SubjectChecker makes this determination
+// based on a set of CollaboratorMatcher objects, which must be added to the
+// SubjectChecker before use.
+type SubjectChecker interface {
+	Allowed(kapi.ObjectReference) error
+	AddMatcher(authorizationapi.CollaboratorMatcher) error
+}
+
+// RoleBindingRestrictions aggregates rules specified by CollaboratorMatcher
+// objects in a namespace and uses these rules to determine whether subjects
+// (users, groups, or service accounts) are collaborators on the namespace.
+type RoleBindingRestrictions struct {
+	SubjectChecker
+
+	oclient oclient.Interface
+	kclient kclientset.Interface
+
+	// namespace is the namespace for which this RoleBindingRestrictions makes
+	// determinations.
+	namespace string
+
+	// userLiterals is a list of literal user names.
+	userLiterals []string
+
+	// userInGroupLiterals is a list of literal group names.
+	userInGroupLiterals []string
+
+	// userSelectors is a list of label selectors.
+	userSelectors []labels.Selector
+
+	// groupLiterals is a list of literal group names.
+	groupLiterals []string
+
+	// groupSelectors is a list of label selectors.
+	groupSelectors []labels.Selector
+
+	// serviceaccountLiterals is a list of tuples of literal service account name
+	// and literal namespace name.
+	serviceaccountLiterals []*authorizationapi.ServiceAccountInNamespace
+
+	// serviceaccountInTheseNamespacesLiterals is a list of literal service
+	// account names.
+	serviceaccountInTheseNamespacesLiterals []string
+}
+
+// NewRoleBindingRestrictions returns a new RoleBindingRestrictions object with
+// no rules.
+func NewRoleBindingRestrictions(ns string, kc kclientset.Interface,
+	oc oclient.Interface) (*RoleBindingRestrictions, error) {
+	return &RoleBindingRestrictions{namespace: ns, kclient: kc, oclient: oc}, nil
+}
+
+// Allowed makes the determination whether a given subject is a collaborator on
+// the namespace.
+func (r *RoleBindingRestrictions) Allowed(subject *kapi.ObjectReference) error {
+	switch subject.Kind {
+	case authorizationapi.SystemUserKind:
```

Comment:
I'd like to start by managing these inside their respect user and group restrictions, but match them based only on literals.